w3lib/url.py

```diff
@@ -422,7 +422,9 @@ def parse_url(url, encoding=None):
     parsed url)
     """
     if isinstance(url, ParseResult):
-        return url
+        # we need to make sure we have unicode parts
+        # otherwise, things like IDNA encoding may fail on bytes
+        return ParseResult(*(to_unicode(part, encoding) for part in url))
```

Comment:
I wonder if we can remove the performance penalty from Python 3 code (if there is a penalty). 

Also, there is an issue with encodings - this assumes utf8 everywhere which may not be the case. Why is Scrapy urlparse-ing bytes urls? Should it use safe_ParseResult instead?